binary_search_tree/binary_search_tree.py

```diff
@@ -0,0 +1,224 @@
+                            parent_node.left = None
+                        elif node.right is None:
+                            # if node only has left child
+                            parent_node.left = node.left
+                        elif node.left is None:
+                            # if node only has right child
+                            parent_node.left = node.right
+                        else:
+                            # if node have both children
+                            if node.right.left is None:
+                                node.right.left = node.left
+                                parent_node.left = node.right
+                            else:
+                                to_replace = node.right
+                                parent_to_replace = node
+                                while to_replace.left is not None:
+                                    parent_to_replace = to_replace
+                                    to_replace = to_replace.left
+                                parent_node.left.data = to_replace.data
+                                parent_to_replace.left = to_replace.right
+                        return
+                    else:
+                        iterator = iterator.left
+                else:
+                    if iterator.right is None:
+                        raise KeyError('Node does not found')
+                    elif iterator.right.data == data:
+                        node = iterator.right
+                        parent_node = iterator
+                        if node.right is None and node.left is None:
+                            # if node has no child
+                            parent_node.right = None
+                        elif node.right is None:
+                            # if node only has left child
+                            parent_node.right = node.left
+                        elif node.left is None:
+                            # if node only has right child
+                            parent_node.right = node.right
+                        else:
+                            # if node have both chidren
+                            if node.right.left is None:
+                                node.right.left = node.left
+                                parent_node.right = node.right
+                            else:
+                                to_replace = node.right
+                                parent_to_replace = node
+                                while to_replace.left is not None:
+                                    parent_to_replace = to_replace
+                                    to_replace = to_replace.left
+                                parent_node.right.data = to_replace.data
+                                parent_to_replace.left = to_replace.right
+                        return
+                    else:
+                        iterator = iterator.right
+        return
+
+    def search_node(self, data):
```

Comment:
Here also using a recursive technique can reduce the code length also there won't be any need to check for root separately.